### PR TITLE
Topic/kleisli

### PR DESCRIPTION
**Pull request by julien-truffaut**
main points:
*   moved `Kleisli` and `Cokleisli` to data
*   renamed `run` to `runKleisli` and `runCokleisli`
*   renamed type parameter from `M` to `F` for consistency
*   added `Monad` and `Strong` instance hierarchies for `Kleisli`

I also added a test to check implicit `Functor` resolution for `Kleisli`. Do we want to have such a test? If yes, is it the right form? 
What sort of tests would you write for `Kleisli` since it is a wrapper for `Function1` we don't have an `Eq`

### CODE REVIEW

**Comment by non** [comment]
@julien-truffaut If we have `Arbitrary[A]` and `Eq[B]` I don't mind using the following as an approximation of `Eq[A => B]`:

```scala
forAll { (a: A) => B.eqv(f(a), g(a)) }
```

**Comment by non** [comment]
One other question: I noticed you changed `run` to `runKleisli` and `runCokleisli`. What's the rationale there? I feel like `run` is just as descriptive -- I could imagine naming the member `function` since it is a `Function1` instance. Maybe there's something I'm missing?

File: data/src/main/scala/cats/data/Kleisli.scala [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+      fa.map(f)
+  }
+}
+
+sealed abstract class KleisliInstances3 {
+  implicit def kleisliFunctor[F[_]: Functor, A]: Functor[Kleisli[F, A, ?]] = new Functor[Kleisli[F, A, ?]] {
+    def map[B, C](fa: Kleisli[F, A, B])(f: B => C): Kleisli[F, A, C] =
+      fa.map(f)
+  }
+}
```
Comment (ceedubs) [inline data/src/main/scala/cats/data/Kleisli.scala]: Travis build is failing because scalastyle wants a new line here.

Comment (julien-truffaut) [inline data/src/main/scala/cats/data/Kleisli.scala]: arhh do you have an automatic way to detect it?

**Comment by julien-truffaut** [comment]
@non I agree that `run` is descriptive enough. The problem arises when you wrap several new-type-like data structure that all have a single `run` accessor, then you have `x.run.run.run`. In that case, I would prefer to have `x.runKleisli.runProcess.runTask`. I noticed this after using scalaz-stream where both `Process` and `Task` have a `run`method.

Now that I think a bit more about it, you could make the same argument for nested `map` and in that case I don't want `mapKleisli`, `mapList` etc... I am curious what other people think about this. I am completely fine to revert that change.

**Comment by milessabin** [comment]
Why not have both so that people can use `run` where there is no ambiguity and `runKleisli`, `runProcess` etc. where there is.

I guess the same argument could be made for `map` vs. `mapKleisli`, `mapList` etc. ... for some reason I don't feel quite so enthusiastic about the more verbose variants in this case.

**Comment by ceedubs** [comment]
@julien-truffaut now that you mention it, I like the idea of something like `x.runKleisli.runProcess.runTask`. I recently wrote a library at work that uses a couple layers of monad transformers, and I think that I've confused my coworkers with my examples that all seem to call `.run.run`. I think @milessabin offers a good approach of offering both.

**Comment by julien-truffaut** [comment]
I would rather have `run` or `runKleisli` but not both. I understand the logic that people can select the one they prefer but a library cannot satisfy everyone taste. I would argue that it is better to have consistent naming for all users but that's just my point of view.

**Comment by julien-truffaut** [comment]
@non do you propose to write an instance `Eq[A => B]` only for the tests? I think it is a reasonable solution

**Comment by non** [comment]
Yes, I think that's a fine strategy.

**Comment by julien-truffaut** [comment]
It doesn't look like there is a consensus on `run`, I propose that I create a separate issue about it and that I revert back `runKleisli` and `runCokleisli` to `run` for this PR. Is it ok?

**Comment by refried** [comment]
@milessabin I too find `.run.run.run` disorienting.  I also find `.map.(_.map.(_.map(...)))` disorienting, but if they're all coming from `Functor`, I don't suppose there's much you can do.

**Comment by julien-truffaut** [comment]
@refried it is good point `map` is a TC method so it has the same name for every instance but `run` doesn't have a single semantic. So it is really confusing to see a sequence of `run`

**Comment by ceedubs** [comment]
@refried you can always compose the functors instead of using nested map calls; it's more fun! It's pretty much always more verbose though :(

**Comment by refried** [comment]
@ceedubs thanks for the tip, I'll try to remember to consider that next time.

**Comment by julien-truffaut** [comment]
I rollbacked the renaming of run and fixed scalastyle issue

**Comment by non** [comment]
This looks great to me.

:+1:

File: data/src/main/scala/cats/data/Cokleisli.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    c compose this
+}
+
+object Cokleisli extends CokleisliInstances with CokleisliFunctions {
+  def pure[F[_], A, B](x: B): Cokleisli[F, A, B] =
+    Cokleisli(_ => x)
+}
+
+sealed trait CokleisliFunctions {
+  /** creates a [[Kleisli]] from a function */
```
Comment (ceedubs) [inline data/src/main/scala/cats/data/Cokleisli.scala]: I think you mean `[[Cokleisli]]` here.

**Comment by ceedubs** [comment]
After https://github.com/non/cats/pull/141#discussion_r24577646 is fixed, this looks good to merge to me.

File: data/src/main/scala/cats/data/Cokleisli.scala (hunk 2 above)
Comment (julien-truffaut) [inline data/src/main/scala/cats/data/Cokleisli.scala]: damn copy / paste, I will change that

**Comment by ceedubs** [comment]
:+1:

**Pull request merged**